encoding/bytestreamsplit/bytestreamsplit_amd64.s

```diff
@@ -0,0 +1,376 @@
+    SHRQ $8, SI
+    ADDQ BX, DI
+
+    MOVB SI, (DI)
+
+    ADDQ $8, AX
+    INCQ DX
+    CMPQ AX, CX
+    JB loop1x8
+done:
+    RET
+
+// func decodeFloat(dst, src []byte)
+TEXT ·decodeFloat(SB), NOSPLIT, $0-48
+    MOVQ dst_base+0(FP), AX
+    MOVQ dst_len+8(FP), BX
+    MOVQ src_base+24(FP), DX
+
+    MOVQ AX, CX
+    ADDQ BX, CX // end
+    SHRQ $2, BX // len
+
+    CMPQ BX, $0
+    JE done
+
+    CMPB ·decodeFloatHasAVX2(SB), $0
+    JE loop1x4
+
+    CMPQ BX, $8
+    JB loop1x4
+
+    MOVQ CX, DI
+    SUBQ AX, DI
+    SHRQ $5, DI
+    SHLQ $5, DI
+    ADDQ AX, DI
+
+    MOVQ $0xFFFFFFFF, SI
+    MOVQ BX, X5
+    MOVQ SI, X6
+    VMOVDQU shuffle8x4<>(SB), Y0
+    VPBROADCASTD X5, Y2
+    VPBROADCASTD X6, Y3
+    VPMULLD scale8x4<>(SB), Y2, Y2
+    VPADDD offset8x4<>(SB), Y2, Y2
+    VMOVDQU Y3, Y4
+loop8x4:
+    VPGATHERDD Y4, (DX)(Y2*1), Y1
+    VPSHUFB Y0, Y1, Y1
+    VMOVDQU Y1, (AX)
+    VMOVDQU Y3, Y4
+
+    ADDQ $32, AX
+    ADDQ $8, DX
+    CMPQ AX, DI
+    JNE loop8x4
+    VZEROUPPER
+
+    CMPQ AX, CX
+    JE done
+loop1x4:
+    MOVQ DX, DI
+    MOVBLZX (DI), R8
+    ADDQ BX, DI
+    MOVBLZX (DI), R9
+    ADDQ BX, DI
+    MOVBLZX (DI), R10
+    ADDQ BX, DI
+    MOVBLZX (DI), R11
+
+    SHLL $8, R9
+    SHLL $16, R10
+    SHLL $24, R11
+
+    ORL R9, R8
+    ORL R10, R8
+    ORL R11, R8
+
+    MOVL R8, (AX)
+
+    ADDQ $4, AX
+    INCQ DX
+    CMPQ AX, CX
+    JB loop1x4
+done:
+    RET
+
+// func decodeDouble(dst, src []byte)
+TEXT ·decodeDouble(SB), NOSPLIT, $0-48
+    MOVQ dst_base+0(FP), AX
+    MOVQ dst_len+8(FP), BX
+    MOVQ src_base+24(FP), DX
+
+    MOVQ AX, CX
+    ADDQ BX, CX
+    SHRQ $3, BX
+
+    CMPQ BX, $0
+    JE done
+
+    CMPB ·decodeDoubleHasAVX512(SB), $0
+    JE loop1x8
+
+    CMPQ BX, $8
+    JB loop1x8
+
+    MOVQ CX, DI
+    SUBQ AX, DI
+    SHRQ $6, DI
+    SHLQ $6, DI
+    ADDQ AX, DI
+
+    VMOVDQU64 shuffle8x8<>(SB), Z0
+    VPBROADCASTQ BX, Z2
+    VPMULLQ scale8x8<>(SB), Z2, Z2
+loop8x8:
+    KXORQ K1, K1, K1
+    KNOTQ K1, K1
+
+    VPGATHERQQ (DX)(Z2*1), K1, Z1
+    VPERMB Z1, Z0, Z1
+    VMOVDQU64 Z1, (AX)
+
+    ADDQ $64, AX
+    ADDQ $8, DX
+    CMPQ AX, DI
+    JNE loop8x8
+    VZEROUPPER
+
+    CMPQ AX, CX
+    JE done
+loop1x8:
+    MOVQ DX, DI
+    XORQ R12, R12
+
+    MOVBQZX (DI), R8
+    ADDQ BX, DI
+    MOVBQZX (DI), R9
+    ADDQ BX, DI
+    MOVBQZX (DI), R10
+    ADDQ BX, DI
+    MOVBQZX (DI), R11
+    ADDQ BX, DI
+
+    SHLQ $8, R9
+    SHLQ $16, R10
+    SHLQ $24, R11
+
+    ORQ R8, R12
+    ORQ R9, R12
+    ORQ R10, R12
+    ORQ R11, R12
+
+    MOVBQZX (DI), R8
+    ADDQ BX, DI
+    MOVBQZX (DI), R9
+    ADDQ BX, DI
+    MOVBQZX (DI), R10
+    ADDQ BX, DI
+    MOVBQZX (DI), R11
+
+    SHLQ $32, R8
+    SHLQ $40, R9
+    SHLQ $48, R10
+    SHLQ $56, R11
+
+    ORQ R8, R12
+    ORQ R9, R12
+    ORQ R10, R12
+    ORQ R11, R12
+
+    MOVQ R12, (AX)
+
+    ADDQ $8, AX
+    INCQ DX
+    CMPQ AX, CX
+    JB loop1x8
+done:
+    RET
+
+GLOBL scale8x4<>(SB), RODATA|NOPTR, $32
```

Comment:
Can you remind me what the following instructions do ?

Reply:
This is how you declare global variables in the Go assembly.

Reply:
Should we just leave a comment in the code explaining that? IMO the more assembler comments the better, especially if this code will end up being maintained by people who aren't as proficient with assembler

Reply:
This is documented here https://go.dev/doc/asm

I don't think I want to document the assembler syntax, we can refer to the Go documentation for this. I left documentation explaining how the algorithms work in the last commit, let me know if you want more docs.